pkg/models/encrypted_file.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright © 2022 Doppler <support@doppler.com>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package models
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/DopplerHQ/cli/pkg/utils"
+)
+
+const Pbkdf2Rounds = 500000
+const LegacyPbkdf2Rounds = 50000
+
+const Base64EncodingPrefix = "base64"
+const HexEncodingPrefix = "hex"
+
+type EncryptedFile struct {
+	Version    int
+	NumRounds  int
+	Encoding   string
+	Ciphertext string
+}
+
+type FileVersionOptions struct {
+	Version          int
+	HasVersionNumber bool
+	HasEncoding      bool
+	HasNumRounds     bool
+}
+
+var FileVersions = map[int]FileVersionOptions{
+	// just ciphertext
+	1: {Version: 1, HasEncoding: false, HasNumRounds: false, HasVersionNumber: false},
+	// ciphertext and encoding
+	2: {Version: 2, HasEncoding: true, HasNumRounds: false, HasVersionNumber: false},
+	// ciphertext, encoding, and pbkdf2 rounds
+	3: {Version: 3, HasEncoding: true, HasNumRounds: true, HasVersionNumber: false},
+	// ciphertext, encoding, pbkdf2 rounds, and explicit version number
+	4: {Version: 4, HasEncoding: true, HasNumRounds: true, HasVersionNumber: true},
+}
+
+func FileVersion(ciphertext string) (int, error) {
+	cParts := strings.Split(ciphertext, ":")
+	if len(cParts) == 4 {
+		return 4, nil
+	}
+	if len(cParts) == 3 {
+		return 3, nil
+	}
+	if len(cParts) == 2 {
+		return 2, nil
+	}
+	if len(cParts) == 1 {
+		return 1, nil
+	}
+	return -1, errors.New("Invalid ciphertext")
+}
+
+func (options *FileVersionOptions) Parse(ciphertext string) (EncryptedFile, error) {
+	var version string
+	var encoding string
+	var numRounds string
+	var data string
+
+	cParts := strings.SplitN(string(ciphertext), ":", 4)
```

Comment:
Curious: Why make a copy of `ciphertext` with `string`?

Reply:
In a previous iteration `ciphertext` was a byte array, but this is no longer needed